.github/workflows/codeql-analysis.yml

```diff
@@ -0,0 +1,71 @@
+# For most projects, this workflow file will not need changing; you simply need
+# to commit it to your repository.
+#
+# You may wish to alter this file to override the set of languages analyzed,
+# or to provide custom queries or build logic.
+name: "CodeQL"
+
+on:
+  push:
+    branches: [master]
+  pull_request:
+    # The branches below must be a subset of the branches above
+    branches: [master]
+  schedule:
+    - cron: '0 5 * * 2'
+
+jobs:
+  analyze:
+    name: Analyze
+    runs-on: ubuntu-latest
+
+    strategy:
+      fail-fast: false
+      matrix:
+        # Override automatic language detection by changing the below list
+        # Supported options are ['csharp', 'cpp', 'go', 'java', 'javascript', 'python']
+        language: ['python']
+        # Learn more...
+        # https://docs.github.com/en/github/finding-security-vulnerabilities-and-errors-in-your-code/configuring-code-scanning#overriding-automatic-language-detection
+
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+      with:
+        # We must fetch at least the immediate parents so that if this is
+        # a pull request then we can checkout the head.
+        fetch-depth: 2
+
+    # If this run was triggered by a pull request event, then checkout
+    # the head of the pull request instead of the merge commit.
+    - run: git checkout HEAD^2
+      if: ${{ github.event_name == 'pull_request' }}
+
+    # Initializes the CodeQL tools for scanning.
+    - name: Initialize CodeQL
+      uses: github/codeql-action/init@v1
+      with:
+        languages: ${{ matrix.language }}
+        # If you wish to specify custom queries, you can do so here or in a config file.
+        # By default, queries listed here will override any specified in a config file. 
+        # Prefix the list here with "+" to use these queries and those in the config file.
+        # queries: ./path/to/local/query, your-org/your-repo/queries@main
```

Comment:
We should add a config file here.
```yml
with:
  languages: ${{ matrix.language }}
  config-file: ./.github/codeql/codeql-config.yml
```

**codeql-config.yml**
```yml
paths:
  - src/streamlink/
  - src/streamlink_cli/
  - tests/
  - docs/
paths-ignore:
  - src/streamlink/_version.py
  - src/streamlink/packages/
  - src/streamlink_cli/packages/
  - 'docs/_themes/**/static/'
```
https://docs.github.com/en/free-pro-team@latest/github/finding-security-vulnerabilities-and-errors-in-your-code/configuring-code-scanning#specifying-directories-to-scan

Otherwise, it'll read and analyse everything, which is not what we want.

Reply:
And if you want to run more CodeQL queries than the default ones, there are also options for this, which might be useful, not sure.
https://docs.github.com/en/free-pro-team@latest/github/finding-security-vulnerabilities-and-errors-in-your-code/configuring-code-scanning#running-additional-queries

> | Query suite | Description |
> | --- | --- |
> | `security-extended` | Queries of lower severity and precision than the default queries |
> | `security-and-quality` | Queries from security-extended, plus maintainability and reliability queries |

Reply:
@gravyboat I've run a code-scan workflow on my test repo and the default queries by Github currently show fewer errors/warnings than lgtm.com (ignoring the files which I put on the ignore list). I think that's because these errors/warnings are code-quality related, which are not enabled by default. And the regular linter doesn't have rules for this either. After adding
```yml
queries: +security-and-quality
```
to the `with:` block of the `github/codeql-action/init@v1` action, it finds the remaining errors/warnings and more.

However, it doesn't seem to respect the `paths-ignore` list, as it lists one error in `src/streamlink/_version.py`.

The run time difference between the query-sets was 6m17s vs 9m11s, assuming the same speed of the CI runner, etc.